server/src/main/java/org/elasticsearch/action/search/SearchProgressListener.java

```diff
@@ -0,0 +1,177 @@
+     * @param shardIndex The index of the shard in the list provided by {@link SearchProgressListener#onListShards(List, boolean)})}.
+     */
+    public void onFetchResult(int shardIndex) {}
+
+    /**
+     * Executed when a shard reports a fetch failure.
+     *
+     * @param shardIndex The index of the shard in the list provided by {@link SearchProgressListener#onListShards(List, boolean)})}.
+     * @param exc The cause of the failure.
+     */
+    public void onFetchFailure(int shardIndex, Exception exc) {}
+
+    final void notifyListShards(List<SearchShard> shards, boolean fetchPhase) {
+        this.shards = shards;
+        try {
+            onListShards(shards, fetchPhase);
+        } catch (Exception e) {
+            logger.warn(() -> new ParameterizedMessage("Failed to execute progress listener on list shards"), e);
+        }
+    }
+
+    final void notifyQueryResult(int shardIndex) {
+        try {
+            onQueryResult(shardIndex);
+        } catch (Exception e) {
+            logger.warn(() -> new ParameterizedMessage("[{}] Failed to execute progress listener on query result",
+                shards.get(shardIndex)), e);
+        }
+    }
+
+    final void notifyQueryFailure(int shardIndex, Exception exc) {
+        try {
+            onQueryFailure(shardIndex, exc);
+        } catch (Exception e) {
+            logger.warn(() -> new ParameterizedMessage("[{}] Failed to execute progress listener on query failure",
+                shards.get(shardIndex)), e);
+        }
+    }
+
+    final void notifyPartialReduce(List<SearchShard> shards, TotalHits totalHits, InternalAggregations aggs, int version) {
+        try {
+            onPartialReduce(shards, totalHits, aggs, version);
+        } catch (Exception e) {
+            logger.warn(() -> new ParameterizedMessage("Failed to execute progress listener on partial reduce"), e);
+        }
+    }
+
+    final void notifyReduce(List<SearchShard> shards, TotalHits totalHits, InternalAggregations aggs) {
+        try {
+            onReduce(shards, totalHits, aggs);
+        } catch (Exception e) {
+            logger.warn(() -> new ParameterizedMessage("Failed to execute progress listener on reduce"), e);
+        }
+    }
+
+    final void notifyFetchResult(int shardIndex) {
+        try {
+            onFetchResult(shardIndex);
+        } catch (Exception e) {
+            logger.warn(() -> new ParameterizedMessage("[{}] Failed to execute progress listener on fetch result",
+                shards.get(shardIndex)), e);
+        }
+    }
+
+    final void notifyFetchFailure(int shardIndex, Exception exc) {
+        try {
+            onFetchFailure(shardIndex, exc);
+        } catch (Exception e) {
+            logger.warn(() -> new ParameterizedMessage("[{}] Failed to execute progress listener on fetch failure",
+                shards.get(shardIndex)), e);
+        }
+    }
+
+    final List<SearchShard> searchShards(List<? extends SearchPhaseResult> results) {
```

Comment:
I would make these methods static and rename them to buildSearchShards

Reply:
I have opened #53373 to address these comments I left last week (besides the one on concurrency that I need to have a closer look at), plus a few other things.